functional-interfaces-streams/src/FunctionalInterfaces.java

```diff
@@ -0,0 +1,181 @@
+
+        //12. Sort a list of employees (name, age) by age using a lambda Comparator.
+        List<Employee> empList = new ArrayList<>(List.of(
+                new Employee("Peti", 10),
+                new Employee("Jani", 11),
+                new Employee("Géza", 12),
+                new Employee("Juli", 13)
+        ));
+        Comparator<Employee> employeeComparator = (e1, e2) -> Integer.compare(e1.age(), e2.age());
+        empList.sort(employeeComparator);
+
+        //13. From a list of integers, find the maximum odd number, return it as Optional<Integer>.
+        List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        Optional<Integer> result = numbers.stream()
+                .filter(n -> n % 2 != 0)
+                .max(Integer::compare);
+        //.max((n1, n2) -> Integer.compare(n1, n2));
+
+        //14. Given a list of words, group them by their first letter using Collectors.groupingBy.
+        Map<Character, List<String>> groupByFirstChar = words.stream()
+                .collect(Collectors.groupingBy(word -> word.charAt(0)));
+        System.out.println(groupByFirstChar);
+
+        //15. Partition a list of numbers into evens and odds using Collectors.partitioningBy.
+        Map<Boolean, List<Integer>> evenOddMap = numbers.stream()
+                .collect(Collectors.partitioningBy(n -> n % 2 == 0));
+        System.out.println(evenOddMap);
+
+        //16. Define your own functional interface StringTransformer with one method transform(String input). Implement it using a lambda to reverse strings.
+        @FunctionalInterface
+        interface StringTransformer {
+            String transform(String input);
+        }
+        StringTransformer StringReverser = input -> new StringBuilder(input).reverse().toString();
+        System.out.println(StringReverser.transform("HELLO"));
+
+        //17. Given a list of people (name, city, age), group them first by city, then by age group (<30, >=30).
+        List<People> listOfPeople = new ArrayList<>(List.of(
+                new People("Gizi", "Győr", 21),
+                new People("Béla", "Budapest", 10),
+                new People("Józsi", "Győr", 75),
+                new People("Fruzsi", "Budapest", 16),
+                new People("Tamás", "Debrecen", 64)
+        ));
+        Map<String, Map<String, List<People>>> grouped = listOfPeople.stream()
+                .collect(Collectors.groupingBy(
+                        People::city,
+                        //person -> person.getCity(),
+                        Collectors.groupingBy(person -> person.age() > 30 ? "<30" : ">30")
+                ));
+        System.out.println(grouped);
+
+        //18. Use a parallel stream to sum numbers from 1 to 1,000,000 and compare performance with sequential streams.
+        int sequentialSum = IntStream.range(1, 1_000_000).reduce(0, Integer::sum);
+        int parallelSum = IntStream.range(1, 1_000_000).parallel().reduce(0, Integer::sum);
```

Comment:
Good - but what is a parallel stream? please explain on Discord or here in a comment :)